youtube_dl/postprocessor/ffmpeg.py

```diff
@@ -431,21 +431,34 @@ def run(self, information):
 
 
 class FFmpegMetadataPP(FFmpegPostProcessor):
+    def __init__(self, downloader=None, preferredinfo=None):
+        super(FFmpegMetadataPP, self).__init__(downloader)
+        if not (isinstance(preferredinfo, dict) or preferredinfo == None):
+            raise TypeError('preferredinfo must be a dictionary, if provided')
```

Comment:
Have you read previous comment at all?

Reply:
@dstftw I'm following an observed pattern (lines 352-355 and 586-589) and raising an exception on the input when necessary. Can you provide a pattern that you're expecting me to follow for this if these examples within the existing code base is not what you're looking for?

`class FFmpegVideoConvertorPP(FFmpegPostProcessor):
    def __init__(self, downloader=None, preferedformat=None):
        super(FFmpegVideoConvertorPP, self).__init__(downloader)
        self._preferedformat = preferedformat`

`class FFmpegSubtitlesConvertorPP(FFmpegPostProcessor):
    def __init__(self, downloader=None, format=None):
        super(FFmpegSubtitlesConvertorPP, self).__init__(downloader)
        self.format = format`

Or are you looking for unit tests to be added for this method?

Reply:
https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement

Reply:
@dstftw assertions are only executed during debug and not during runtime. I added an exception error so that it is in effect 100% of time. Are you saying that your preference is that errors are not thrown consistently?

From the documentation you provided, assert is equivalent to:
`if __debug__:
   if not expression: raise AssertionError`

Reply:
I guess you misunderstand the whole point of assertions. You should distinguish cases when data comes externally and internally. In this case the latter takes place. The input data is build by our own code internally therefore we ourselves are responsible for guaranteeing its validity so that it's pointless to always perform this runtime check. If it turns out to be invalid then it's a bug in our code that is idiomatically as precondition covered by assert in debug.

Reply:
@dstftw A. I'm not seeing assertions in ffmpeg at all, so this pattern is not followed and B. I'm not seeing anywhere that datatypes are validated. This is an argument passed in externally. There isn't a pattern at this validation happens further up in the service. I'm happy to swap it out for an assert, but without upstream validation, it is an assert on an external data element

Reply:
Again: asserts are for **invariant checks** during **development**. Once you're done you may remove them freely that's why there are almost no asserts in code base. The point is that there is no need to throw in such cases. If you want to check while developing add an assert. If you're sure enough you won't stumble while developing don't use asserts at all.